rmw_cyclonedds_cpp/src/serdata.cpp

```
@@ -685,11 +749,351 @@ static std::string get_type_name(const char * type_support_identifier, void * ty
+static void dynamic_type_add_array(
+  dds_dynamic_type * dstruct, dds_entity_t dds_ppant, const MemberType * member, const dds_dynamic_type_t ddt)
+{
+  dds_dynamic_type_t dseq;
+
+  if (member->array_size_) 
+  { 
+    if (!member->is_upper_bound_)
+    {
+      dseq = dds_dynamic_type_create(dds_ppant, 
+       get_dynamic_type_descriptor(DDS_DYNAMIC_ARRAY, member->name_, 1, (uint32_t*)(&member->array_size_), ddt));
+    } else {
+      dseq = dds_dynamic_type_create(dds_ppant, 
+        get_dynamic_type_descriptor(DDS_DYNAMIC_SEQUENCE, member->name_, 1, (uint32_t*)(&member->array_size_), ddt));
+    }
+  } else {
+    dseq = dds_dynamic_type_create(dds_ppant,
+     get_dynamic_type_descriptor(DDS_DYNAMIC_SEQUENCE, member->name_, 0, nullptr, ddt)); 
+  }
+
+  dds_return_t ret = dds_dynamic_type_add_member(dstruct, get_dynamic_member_descriptor(dseq, member->name_));
+  if (ret != DDS_RETCODE_OK)
+    RCUTILS_LOG_ERROR("failed to add type member sequence/array with dynamic type");
+}
+
+template<typename MemberType>
+static void dynamic_type_add_member(
+    dds_dynamic_type_t * dstruct, dds_entity_t dds_ppant, const MemberType * member, const dds_dynamic_type_kind_t type)
+{
+  dds_return_t ret;
+  if (type == rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING ||
+      type == rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE)
+    return;
+
+  if (!member->is_array_)
+  {
+    ret = dds_dynamic_type_add_member(dstruct, get_dynamic_member_descriptor_prim(type, member->name_));
+    if (ret != DDS_RETCODE_OK)
+      RCUTILS_LOG_ERROR("failed to add type member: %d", type);
+  } else {
+    dynamic_type_add_array(dstruct, dds_ppant, member, type);
+  }
+}
+
+static void dynamic_type_register(struct sertype_rmw * st, dds_dynamic_type_t dt, dds_entity_t dds_ppant)
+{
+  dds_typeinfo_t *type_info;
+  auto rc = dds_dynamic_type_register(&dt, &type_info);
+  if (rc != DDS_RETCODE_OK)
+    RMW_SET_ERROR_MSG("dds_dynamic_type_register failed to register type");
```

It would be better to propagate the error out here, I think. I suspect continuing after an error will crash ... (This applies in a few more places.)

---

I agree and change all the same moments to `assert`.